### PR TITLE
Add Postgres to metrics collector

### PR DESCRIPTION
**Pull request by ZuGao**
## Summary

* Add Postgres to metrics collector.
* Refactoring of the already existing object storage metrics collector.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,6 +62,52 @@ $ cd appuio-cloud-reporting
 $ make docker-compose-up
 ```
 
+For the first time you start the database locally, check `Local Development` on the Readme in [appuio-cloud-reporting](https://github.com/appuio/appuio-cloud-reporting/blob/master/README.md#local-development) or follow these steps:
+* Next command asks for a password, it's "reporting":
+```
+S createdb --username=reporting -h localhost -p 5432 appuio-cloud-reporting-test
```
Comment (ZuGao) [inline README.md:68]:
```suggestion
$ createdb --username=reporting -h localhost -p 5432 appuio-cloud-reporting-test
```

**Review by ZuGao — COMMENTED** [review]
Checked only the docs

File: README.md [hunk 2]
```diff
@@ -61,6 +62,52 @@ $ cd appuio-cloud-reporting
...
+      timeOfDay: "12:00:00"
+    size:
+      plan: hobbyist-2
+    service:
+      majorVersion: "14"
+  writeConnectionSecretToRef:
+    name: postgres-connection-details
+```
+
+Once the database is created and `Ready`, you can run locally the command:
```
[the remark this answers is not in the document]
Comment (susana-garcia) [reply README.md:102]: yes, that was already in the readme (line 30-34)

File: .codeclimate.yml [hunk 3]
```diff
@@ -1,4 +1,7 @@
 version: "2"
+checks:
+  similar-code:
+    enabled: false
```
Comment (mweibel) [inline .codeclimate.yml:4]: in [cloudscale-metrics-collector](https://github.com/vshn/cloudscale-metrics-collector/pull/21#pullrequestreview-1183842033) @ccremer mentioned that I should remove all codeclimate stuff. Not sure if that's a general rule for vshn repos but something worth considering.

File: pkg/database/database.go [hunk 4]
```diff
@@ -14,238 +15,167 @@ import (
...
+// rollback rolls back transaction in case of error in previous steps
+func (d *Database) rollback(dctx *Context, err error) error {
+	log := ctrl.LoggerFrom(dctx)
 	if err != nil {
-		return fmt.Errorf("cannot commit transaction for initial database configuration: %w", err)
+		log.Error(err, "found error in pipeline")
+		err = dctx.transaction.Rollback()
+		if err != nil {
+			return fmt.Errorf("cannot rollback transaction: %w", err)
+		}
```
Comment (mweibel) [inline pkg/database/database.go:159]: if there was an error and the rollback was successful, this would return no error and possibly exit code 0. Is that intentional?

File: pkg/database/dbaas.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	return p.RunWithContext(dctx)
+}
+
+// saveUsageToDatabase saves each previously aggregated DBaaS to the billing database
+func (d *DBaaSDatabase) saveUsageToDatabase(dctx *Context) error {
+	log := ctrl.LoggerFrom(dctx)
+	for _, aggregated := range *dctx.AggregatedObjects {
+		err := d.ensureAggregatedServiceUsage(dctx, aggregated)
+		if err != nil {
+			log.Error(err, "Cannot save aggregated DBaaS service record to billing database")
```
Comment (mweibel) [inline pkg/database/dbaas.go:39]: should we add some log info about which part failed or does `err` have that already?

File: pkg/database/postgres.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+type productDBaaS struct {
+	Plan   string
+	Target string
+	Amount float64
+}
+
+var (
+	// Hobbyist2 plan
+	Hobbyist2 = productDBaaS{Plan: "hobbyist-2", Target: "1411", Amount: 0.05811000}
```
Comment (mweibel) [inline pkg/database/postgres.go:17]: [pricing](https://products.docs.vshn.ch/products/appcat/exoscale_dbaas.html#_pricing) mentions a different price for hobbyist 2 (I only checked this one, didn't check other ones). 

Which one is correct? Or are both correct since there's some margin applied later on?

File: main.go [hunk 7]
```diff
@@ -67,7 +65,8 @@ func newApp() (context.Context, context.CancelFunc, *cli.App) {
 			},
 		},
 		Commands: []*cli.Command{
```
Comment (mweibel) [inline main.go:67]: nit: would it not be possible to add global flags like exoscale API keys and k8s connection creds to the global app flags instead of adding them individually by command?

File: pkg/database/postgres.go (hunk 6 above)
Comment (susana-garcia) [reply pkg/database/postgres.go:17]: good question, and I'm not sure about it, maybe @ZuGao knows?

File: .codeclimate.yml (hunk 3 above)
Comment (ZuGao) [reply .codeclimate.yml:4]: Cloudscale collector was developed a bit differently than usual go code. I think that was the main reason why we had to remove it. We had some issue with code climate here as well and we had to adjust a bit the config as we thought that we should add more tests and further refactoring when we will merge both metrics collectors.

File: pkg/service/dbaas/service.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+// Context is the context of the DBaaS service
+type Context struct {
+	context.Context
+	dbaasDetails     []Detail
+	exoscaleDBaasS   []*egoscale.DatabaseService
+	aggregatedDBaasS map[db.Key]db.Aggregated
+}
+
+// Service is the context of the DBaaS service
```
Comment (glrf) [inline pkg/service/dbaas/service.go:44]: Nitpick: Which one is the context now? :sweat_smile:

File: pkg/database/postgres.go (hunk 6 above)
Comment (ZuGao) [reply pkg/database/postgres.go:17]: ah that's interesting. I think I got the prices from one of my colleagues on Friday which might be wrong as I'm checking now. @Kidswiss can you please check and let us know here what prices are correct?

File: pkg/service/dbaas/service.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	p.WithSteps(
+		p.NewStep("Fetch cluster managed DBaaS", s.fetchManagedDBaaS),
+		p.NewStep("Fetch exoscale DBaaS usage", s.fetchDBaaSUsage),
+		p.NewStep("Filter supported DBaaS", filterSupportedServiceUsage),
+		p.NewStep("Aggregate DBaaS services by namespace and plan", aggregateDBaaS),
+		p.WithNestedSteps("Save to billing database", hasAggregatedInstances,
+			p.NewStep("Get billing date", s.getBillingDate),
+			p.NewStep("Save to database", s.saveToDatabase),
+		),
+	)
```
Comment (glrf) [inline pkg/service/dbaas/service.go:79]: I know it's already all over the codebase so we don't (shouldn't) change this now, but I honestly think these Pipelines just make everything harder to read with no clear benefit..

File: pkg/service/dbaas/service.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+		}
+
+		for _, managedResource := range managedResources.Items {
+			dbaasDetail := Detail{
+				DBName: managedResource.GetName(),
+			}
+			if organization, exist := managedResource.GetLabels()[service.OrganizationLabel]; exist {
+				dbaasDetail.Organization = organization
+			} else {
+				// cannot get organization from bucket
```
Comment (glrf) [inline pkg/service/dbaas/service.go:102]: Nitpick: We're not handling buckets anymore :)

File: pkg/database/dbaas.go (hunk 5 above)
Comment (ZuGao) [reply pkg/database/dbaas.go:39]: the error is supposed to have the details what failed in the pipeline. So we might need to check in the underlying pipeline if we catch every relevant error.

File: pkg/database/database.go (hunk 4 above)
Comment (ZuGao) [reply pkg/database/database.go:159]: I think it's a potential bug here yes. This method is being used not only for aggregated objects but for ensuring the initial billing database configuration. At this point we might stop the application altogether even for single aggregate objects. We might miss as you said object not being saved in DB and no one would notice it in the logs.

File: .codeclimate.yml (hunk 3 above)
Comment (ccremer) [reply .codeclimate.yml:4]: I would remove codeclimate whenever you have some time for repo cleanup when you touch a Go repo at VSHN. In our bootstrap repo we also replaced codeclimate already.

> I think that was the main reason why we had to remove it.

No, we're replacing codeclimate because it seems to abandon the Go language support.

Comment (ZuGao) [reply .codeclimate.yml:4]: > No, we're replacing codeclimate because it seems to abandon the Go language support.

Is there an official page for that?

File: pkg/service/dbaas/service.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+	for _, groupVersionResource := range groupVersionResources {
+		managedResources, err := s.k8sClient.Resource(groupVersionResource).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot list managed resource %s from cluster: %w", groupVersionResource.Resource, err)
+		}
+
+		for _, managedResource := range managedResources.Items {
+			dbaasDetail := Detail{
+				DBName: managedResource.GetName(),
+			}
```
Comment (glrf) [inline pkg/service/dbaas/service.go:98]: What happens if there are two databases with the same name (msql "foo" and postgres "foo")? Are DBaaS names unique across database types on exoscale?

And even if they are this is a potential problem.  Let's say tenant A has a mysql resource called "foo" and tenant B has a (not ready) postgres resource. We'll potentially bill tenant B for the mysql server right?

I know this is technically not an issue right now but If you start to implement this in a generic way it should actually also work when we handle multiple DBaaS

File: main.go (hunk 7 above)
Comment (ccremer) [reply main.go:67]: Yes, it's possible to add global flags. However, the way the CLI handles global flags (they have to be provided before the subcommand, e.g. `binary --global-flag subcommand`), I recommend to put each flag in a constructor/func like this: https://github.com/vshn/provider-exoscale/blob/master/operator_command.go#L34 and then use the func to provide the flag for each subcommand.
That way, you can do the following:
`binary subcommand --global-flag`
It allows still being able to control every flag for a subcommand (e.g. for some subcommands you might want to hide/disable certain "global" flags)

File: pkg/service/dbaas/service.go (hunk 9 above)
Comment (susana-garcia) [reply pkg/service/dbaas/service.go:79]: I agree, it's not very go-like style, but it's hard to change now.

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (susana-garcia) [reply pkg/service/dbaas/service.go:98]: In the first loop we use `GroupVersionResource` to request k8s resources by type, so in every iteration we only handle one type of database (postgres, mysql...)
I don't think the name is unique in exoscale, but when we save the information in the db, we don't even use it, we just take into account namespace, organization, database type and plan. So we can basically delete the name completely from here.
Thoughts?

File: pkg/service/dbaas/service.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+	for _, dbaasDetail := range ctx.dbaasDetails {
+		log.V(1).Info("Checking DBaaS", "instance", dbaasDetail.DBName)
+
+		if dbaasUsage, exists := dbaasServiceUsageMap[dbaasDetail.DBName]; exists {
+			log.V(1).Info("Found exoscale dbaas usage", "instance", dbaasUsage.Name, "instance created", dbaasUsage.CreatedAt)
+			key := db.NewKey(dbaasDetail.Namespace, *dbaasUsage.Plan, *dbaasUsage.Type)
+			aggregated := aggregatedDBaasS[key]
+			aggregated.Key = key
+			aggregated.Organization = dbaasDetail.Organization
+			aggregated.Value++
```
[the remark this answers is not in the document]
Comment (susana-garcia) [reply pkg/service/dbaas/service.go:183]: yes, that's correct.

File: pkg/database/dbaas.go (hunk 5 above)
Comment (susana-garcia) [reply pkg/database/dbaas.go:39]: the error has it already. So for example, if the app fails in step `Save facts`, the log will look like:
```
...
ERROR | exoscale-metrics-collector.dbaas | database/dbaas.go:39 Cannot save aggregated DBaaS service record to billing database | {"error": "step 'Saving DBaaS usage namespace default' failed: error found in pipeline: step 'Save facts' failed: some error"}
...
```

File: main.go (hunk 7 above)
Comment (susana-garcia) [reply main.go:67]: yes, I believe, the way @ccremer mentioned is how it's done for the current commands. Unless I'm missing something else. Feel free to reopen the conversation again, if so.
Maybe you meant to add them as Flags on the `cli.App`? In that case, we'll need to read them from the context in the sub commands and I don't think we want that.

File: pkg/database/postgres.go (hunk 6 above)
Comment (susana-garcia) [reply pkg/database/postgres.go:17]: FYI: the prices we use come from the [Portal Exoscale API](https://portal.exoscale.com/api/pricing/dbaas-pg)
What I saw is that we didn't add plan `hobbyist-1`, can you double checked that it exists in Exoscale? I don't have access to Exoscale, sorry.

File: .codeclimate.yml (hunk 3 above)
Comment (susana-garcia) [reply .codeclimate.yml]: Does anyone know why `codeclimate` is still in the checks of the PR, even though I deleted everything related to it? Do we need to merge to master before it's completely gone or am I missing something?

File: pkg/database/postgres.go (hunk 6 above)
Comment (Kidswiss) [reply pkg/database/postgres.go:17]: To ensure the correct prices is in another ticket: https://ticket.vshn.net/browse/APPCAT-264

For this PR it's okay to have dummy values, but if you want you can already get the prices from the CSV in APPCAT-264.

Comment (susana-garcia) [reply pkg/database/postgres.go:17]: perfect, thank you (that's the ones we used). I'll add `hobbyist-1` that was missing 👍

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (Kidswiss) [reply pkg/service/dbaas/service.go:98]: my 2 cents about the name uniqueness:

The names for the DBaaS instances are the same names as the managed resource on the cluster has. That name is unique for the whole cluster as the MR is cluster scoped and has a random appendix. Given that each cluster should have its own exoscale project, there shouldn't be any collisions.

Even if the exoscale project is shared accross different clusters, the chances for a collision are very minimal due to the random naming.

File: pkg/database/postgres.go (hunk 6 above)
Comment (Kidswiss) [reply pkg/database/postgres.go:19]: @susana-garcia aah, I think you used the wrong column then :)

According to APPCAT-264, the relevant prices are in the "Final Price / hour" column ;)

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (glrf) [reply pkg/service/dbaas/service.go:98]: But that doesn't solve the issue I presented right?  In `aggregateDBaaS` we match usage and details based on the name. Am I completely missing something?

File: pkg/database/postgres.go (hunk 6 above)
Comment (susana-garcia) [reply pkg/database/postgres.go:19]: oh, I see. Sorry, I missed that part on the ticket. I'll update them 😄

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (glrf) [reply pkg/service/dbaas/service.go:98]: > Even if the exoscale project is shared accross different clusters, the chances for a collision are very minimal due to the random naming.

IMO we shouldn't rely on this being unlikely for actually billing resources..

Comment (Kidswiss) [reply pkg/service/dbaas/service.go:98]: > But that doesn't solve the issue I presented right? In aggregateDBaaS we match usage and details based on the name. Am I completely missing something?

True, I haven't checked the code in detail yet, so I'm missing some context what exactly is wrong here.

File: main.go (hunk 7 above)
Comment (mweibel) [reply main.go:67]: as said, this is just a nit. 

To me, it does not make sense to have the same flags added individually to each subcommand, especially not essential ones like exoscale API keys (this is the exoscale metrics collector after all) and k8s url/token (which is required everywhere). 

The order of how to provide the flags to the command is, to me, neglectable since it's not a user facing command, we configure the cronjob once and be done with it. Also IIRC we provide the flags via env anyway so that's not an issue. 

Reading them from the context in the sub command: ok that's a bit inconvenient but neglectable IMO. 

but that's just my 2c on a relatively minor issue and shouldn't prevent the PR from getting merged.

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (susana-garcia) [reply pkg/service/dbaas/service.go:98]: maybe we could use the k8s resource UID to differentiate between them. But, then we might need to add it in the facts somehow 🤔

Comment (glrf) [reply pkg/service/dbaas/service.go:98]: Don't we just need to match exoscale resources and managed resources not just on name but also on DB type?  I think that should completely solve the issue and would only need a very minimal change

File: pkg/database/database.go [hunk 13]
```diff
@@ -14,238 +15,169 @@ import (
...
 	return nil
 }
 
-func (d *Database) adjustStorageSizeUnit(ctx *transactionContext) error {
-	var quantity float64
-	if query.Unit == defaultUnit {
-		quantity = ctx.aggregatedBucket.StorageUsed / 1024 / 1024 / 1024
-	} else {
-		return fmt.Errorf("unknown query unit %s", query.Unit)
+func (d *Database) saveFacts(dctx *Context) error {
```
Comment (Kidswiss) [inline pkg/database/database.go:122]: Due to the aggregation we'll probably lose some data here. We said, that we bill a database for the whole hour if it's in any of the metrics runs.

If we don't aggregate it's very simple, we just write a 1 into the database, if we see an instance at any given hour, easy.

But we aggregate thing here, so if we run, let's say every 10 minutes, it could happen that the customer has 3 running instances for the given hour. 10 minutes later he deleted one instance and the metrics only return 2. We lose the billing for one instance.

So on each run we have to determine for the given hour, if the value that was written before is higher or lower and only write the higher value.

TL;DR: check if there's already a value for the current hour, if the existing value is smaller, overwrite it, if it isn't keep it.

File: pkg/service/dbaas/service.go (hunk 11 above)
Comment (ZuGao) [reply pkg/service/dbaas/service.go:98]: First of all the names of DBaaS are unique across organisations in Exoscale. It means you cannot have a postgres and a mysql instance called "db". You can test it fairly easy in Exoscale DBaaS tab by creating two different types of DBaaS. This fact alone insures that the type is irrelevant for our "key" and that the name is sufficient. Now we have an edge case, as @glrf pointed out, we would count 2 instances twice when the cluster has 2 resources with the same name but different DBaaS type (first in Ready state and second in error naturally because of the name constraint). For this edge case I update the code and now I'm matching not just the name but the the DBaaS type as well.

File: pkg/service/dbaas/service.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ZuGao) [reply pkg/service/dbaas/service.go:183]: just a small correction, we plan to run it every 15 min and it would make sure 1 record exist in the billing database per hour.

[the remark this answers is not in the document]
Comment (susana-garcia) [reply pkg/service/dbaas/service.go:183]: oh, sorry, I forgot about that.

File: pkg/database/database.go (hunk 13 above)
Comment (ZuGao) [reply pkg/database/database.go:122]: I have updated the code and introduced the function `isFactsUpdatable` which will make sure that fact table is updated only for DBaaS service. Also the record will only be saved if the quantity is higher than the previous value. It will never save a lower value. Moreover, we have to update the [library](https://github.com/vshn/cloudscale-metrics-collector/tree/master/pkg) of the billing DB to support updating facts table. It is already [implemented](https://github.com/vshn/cloudscale-metrics-collector/blob/master/pkg/productsmodel/products.go#L75) for products.

Comment (ZuGao) [reply pkg/database/database.go:122]: Created a [PR](https://github.com/vshn/cloudscale-metrics-collector/pull/29) in the library code.

**Review by Kidswiss — APPROVED** [review]

File: pkg/database/dbaas.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+		Plan:         plan,
+	}
+
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.WithNestedSteps(fmt.Sprintf("Saving DBaaS usage namespace %s", namespace), nil,
+			p.NewStep("Begin database transaction", d.beginTransaction),
+			p.NewStep("Ensure necessary models", d.ensureModels),
+			p.NewStep("Get best match", d.getBestMatch),
+			p.WithNestedSteps("Save to billing database", isFactsUpdatable,
```
Comment (Kidswiss) [inline pkg/database/dbaas.go:85]: Just to be sure, the nested step is only executed if the first one returns true?

Comment (ZuGao) [reply pkg/database/dbaas.go:85]: yes

**Review by glrf — APPROVED** [review]
LGTM now :rocket:

File: pkg/service/dbaas/service.go [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+
+// aggregateDBaaS aggregates DBaaS services by namespaces and plan
+func aggregateDBaaS(ctx *Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Aggregating DBaaS instances by namespace and plan")
+
+	dbaasServiceUsageMap := make(map[string]egoscale.DatabaseService, len(ctx.exoscaleDBaasS))
+	for _, usage := range ctx.exoscaleDBaasS {
+		dbaasServiceUsageMap[*usage.Name] = *usage
+	}
```
[the remark this answers is not in the document]
Comment (ZuGao) [reply pkg/service/dbaas/service.go:172]: Great suggestion. I'm adding it now.
